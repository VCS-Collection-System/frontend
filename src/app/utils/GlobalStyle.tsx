import * as React from 'react';
import { createGlobalStyle } from 'styled-components';

const GlobalStyle = () => {

  const fontColor = "rgb(180, 180, 178)"; 
  const backgroundColor200 = "var(--pf-global--BackgroundColor--dark-200)";
  const backgroundColor300 = "var(--pf-global--BackgroundColor--dark-300)";
  const backgroundColor400 = "var(--pf-global--BackgroundColor--dark-400)"; 
  const borderColor200 = "var(--pf-global--BorderColor--200)"; 

  const GlobalStyles = createGlobalStyle`
    .pf-c-page__main {
      background-color: ${backgroundColor400}
    }
    .pf-c-button.pf-m-plain#span {
      color: ${fontColor};
    }
    .pf-c-title.pf-m-3xl { 
      color: ${fontColor};
    }
    .pf-c-card {
      --pf-c-card--BackgroundColor: ${backgroundColor200};
      color: ${fontColor};
    }
    .pf-c-card .pf-c-card{
      --pf-c-card--BackgroundColor: ${backgroundColor400};
      color: ${fontColor};
    }
    .pf-c-card .pf-c-card .pf-c-card{
      --pf-c-card--BackgroundColor: ${backgroundColor200};
      color: ${fontColor};
    }
    .pf-c-form-control:disabled {
      --pf-c-form-control--BackgroundColor: ${backgroundColor300};
      color: ${fontColor};
    };
    .pf-c-form-control {
      --pf-c-form-control--BackgroundColor: ${backgroundColor300};
      --pf-c-form-control--BorderTopColor: ${backgroundColor300};
      --pf-c-form-control--BorderRightColor: ${backgroundColor300};
      --pf-c-form-control--BorderBottomColor: ${backgroundColor300};
      --pf-c-form-control--BorderLeftColor: ${backgroundColor300};
      color: ${fontColor};
    };
    .pf-c-button.pf-m-tertiary {
      color: ${fontColor};
    };
    .pf-c-button.pf-m-control {
      background-color: ${backgroundColor300};
      --pf-c-button--after--BorderColor: ${backgroundColor300};
      color: ${fontColor};
    };
    .pf-c-calendar-month {
      --pf-c-calendar-month--BackgroundColor: ${backgroundColor300};
      color: ${fontColor};
    };
    .pf-c-calendar-month__dates-cell::before {
      background-color: ${backgroundColor300};
      color: ${fontColor};
    };
    .pf-c-calendar-month__date {
      color: ${fontColor};
    };
    .pf-c-select__toggle {
      background-color: ${backgroundColor300};
      color: ${fontColor};
    };
    .pf-c-form-control[readonly] {
      background-color: ${backgroundColor300};
      color: ${fontColor};
    };
    .pf-c-form__helper-text {
      color: ${fontColor};
    };
    .pf-c-menu {
      --pf-c-menu--BackgroundColor: ${backgroundColor300};
      color: color: ${fontColor};
    };
    .pf-c-menu__item {
      background-color: ${backgroundColor300};
      color: ${fontColor};
    }
    .pf-c-input-group {
      --pf-c-input-group--BackgroundColor: ${backgroundColor300};
    }
    .pf-c-search-input {
      --pf-c-search-input__text--before--BorderColor: ${borderColor200};
      --pf-c-search-input__text--after--BorderBottomColor: ${borderColor200};
      color: ${fontColor};
    }
    .pf-c-search-input__text-input {
      color: ${fontColor};
    }
    .pf-c-modal-box {
      --pf-c-modal-box--BackgroundColor: ${backgroundColor300};
      color: ${fontColor};
    }
    .pf-c-radio {
      --pf-c-radio__label--Color: ${fontColor};
    }
    .pf-c-button.pf-m-plain {
      color: ${fontColor};
    }
    `;

    return <GlobalStyles />
};

export default GlobalStyle;