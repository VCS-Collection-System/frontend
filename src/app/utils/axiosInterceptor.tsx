import axios from "axios";

const axiosInterceptor = axios.create({})

axiosInterceptor.interceptors.request.use(
    config => {
        config.headers['Authorization'] = 'Bearer ' + sessionStorage.getItem("token");
        return config;
    },
    error => {
        Promise.reject(error)
    });
export default axiosInterceptor;