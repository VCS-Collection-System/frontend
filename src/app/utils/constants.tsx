declare global {  
  interface Window {
    REACT_APP_PAM_URL_BASE: string;
    REACT_APP_PAM_CONTAINER_ID: string;
    REACT_APP_KEYCLOAK_COUNTRY: string;
    REACT_APP_KEYCLOAK_EMAIL: string;
    REACT_APP_KEYCLOAK_FIRST_NAME: string;
    REACT_APP_KEYCLOAK_ID: string;
    REACT_APP_KEYCLOAK_LAST_NAME: string;
    REACT_APP_MAINTENANCE_BANNER: boolean;
    REACT_APP_MAINTENANCE_BANNER_TEXT: string;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    REACT_APP_FEATURE_FLAGS: Array<Record<string, any>>;
    REACT_APP_MAX_DOSES: Record<string, number>;
    REACT_APP_CONSENT_LABEL: string;
    REACT_APP_ALTERNATE_CONSENT_LABEL: string;
    REACT_APP_BLOCKED_COUNTRIES: Array<string>;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    REACT_APP_GLOBAL_CONFIG: Record<string, any>;
  }
}

// PAM urls
export const PAM_BASE_URL = window.REACT_APP_PAM_URL_BASE; 
export const SUBMIT_VAX_URL = PAM_BASE_URL + '/attestation/vax';
export const SUBMIT_TEST_URL = PAM_BASE_URL + '/attestation/covid-test-result';
export const PAM_TASK_URL = PAM_BASE_URL + '/tasks/';
export const RESERVED_TASK_URL = PAM_TASK_URL + 'owners?status=Reserved&user=';
export const INPROGRESS_TASK_URL = PAM_TASK_URL + 'owners?status=InProgress&user=';
export const ALL_TASK_URL = PAM_TASK_URL + 'owners?status=Ready&pageSize=200';
export const PRIORITY_URL = PAM_BASE_URL + '/query/priority-results';
export const PRIORITY_CONFIRM_URL = PAM_BASE_URL + '/query/priority-result/';
export const ATTACHMENT_URL = PAM_BASE_URL + '/attestation/attachment';
export const EMPLOYEE_URL = PAM_BASE_URL + '/query/employee/';
export const EMPLOYEE_NAME_URL = PAM_BASE_URL + '/query/employee-name/';
export const EMPLOYEE_VAX_HISTORY_URL = PAM_BASE_URL + '/query/vax-document/';
export const EMPLOYEE_TEST_HISTORY_URL = PAM_BASE_URL + '/query/test-result-document/';
export const VAX_TASK_DOCUMENT_URL = PAM_BASE_URL + '/query/task/document/';
export const DELETE_VAX_URL = PAM_BASE_URL + '/attestation/delete-vax-record/';
export const CONFIG_URL = PAM_BASE_URL + '/query/country-configs/';
export const SAVE_CONFIG_URL = PAM_BASE_URL + '/attestation/country-configs';

// List of countries allow to use this app
export const BLOCKED_COUNTRIES = window.REACT_APP_BLOCKED_COUNTRIES || []

export const KONAMI_CODE = [ "ArrowUp", "ArrowUp", "ArrowDown", "ArrowDown", "ArrowLeft", "ArrowRight", "ArrowLeft", "ArrowRight", "KeyB", "KeyA"];

// Maintenance Banner
export const MAINTENANCE_BANNER = window.REACT_APP_MAINTENANCE_BANNER || false
export const MAINTENANCE_BANNER_TEXT = window.REACT_APP_MAINTENANCE_BANNER_TEXT || ""

// Keycloak claims
export const KEYCLOAK_COUNTRY = window.REACT_APP_KEYCLOAK_COUNTRY
export const KEYCLOAK_EMAIL = window.REACT_APP_KEYCLOAK_EMAIL
export const KEYCLOAK_FIRST_NAME = window.REACT_APP_KEYCLOAK_FIRST_NAME
export const KEYCLOAK_ID = window.REACT_APP_KEYCLOAK_ID
export const KEYCLOAK_LAST_NAME = window.REACT_APP_KEYCLOAK_LAST_NAME
export const FEATURE_FLAGS = window.REACT_APP_FEATURE_FLAGS || []

// Maximum doses allowed by vaccine manufacturer
export const MAX_DOSES = window.REACT_APP_MAX_DOSES || {}

export const CONSENT_LABEL = window.REACT_APP_CONSENT_LABEL
export const ALTERNATE_CONSENT_LABEL = window.REACT_APP_ALTERNATE_CONSENT_LABEL

// Global country config
export const GLOBAL_CONFIG = window.REACT_APP_GLOBAL_CONFIG || {};

