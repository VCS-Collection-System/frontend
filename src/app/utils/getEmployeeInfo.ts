import {
  KEYCLOAK_FIRST_NAME, 
  KEYCLOAK_LAST_NAME, 
  KEYCLOAK_EMAIL,
  KEYCLOAK_COUNTRY,
  KEYCLOAK_ID
} from "@app/utils/constants"; 

export function getEmployeeInfo(keycloak) {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const jp = require('jsonpath');
  const profile = keycloak.profile;
  const employeeFirstName = jp.query(profile, KEYCLOAK_FIRST_NAME)[0];
  const employeeLastName = jp.query(profile, KEYCLOAK_LAST_NAME)[0];
  const employeeEmail = jp.query(profile, KEYCLOAK_EMAIL)[0];
  const employeeCountry = jp.query(profile, KEYCLOAK_COUNTRY)[0];
  const employeeID = jp.query(profile, KEYCLOAK_ID)[0];

  const isApprover = () => {
    if(employeeCountry && 
      (keycloak.hasRealmRole("vcs-superapprover") || keycloak.hasRealmRole("vcs-approver-" + employeeCountry.toLowerCase()))){
        return true;
    }

    return false;
  }

  const isSuperApprover = () => {
    return keycloak.hasRealmRole("vcs-superapprover");
  }

  return {
    firstName: employeeFirstName, 
    lastName: employeeLastName,
    email: employeeEmail,
    country: employeeCountry,
    id: employeeID,
    isApprover: isApprover(),
    isSuperApprover: isSuperApprover()
  }
}
