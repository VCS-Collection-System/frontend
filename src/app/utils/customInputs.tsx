import { useEffect, useState } from "react";
import { KONAMI_CODE } from './constants'

const useInputEvent = () => {
  const [key, setKey] = useState(null);
  useEffect(() => {
    const keyDownHandler = ({ code }) => setKey(code);
    const keyUpHandler = () => setKey(null);
    window.addEventListener('keydown', keyDownHandler);
    window.addEventListener('keyup', keyUpHandler);
    return () => {
      window.removeEventListener("keydown", keyDownHandler);
      window.removeEventListener("keyup", keyUpHandler)
    }
  }, []);
  return key;
}

const useSecretCode = (secretCode) => {
  const [count, setCount] = useState(0);
  const [success, setSuccess] = useState(false);
  const key = useInputEvent();
  
  useEffect(() => {
    if (key == null) return;
    if (key !== secretCode[count]) {
      setCount(0);
      return;
    }
    
    setCount((state) => state + 1);
    if (count + 1 === secretCode.length) {
      setSuccess(!success);
      setCount(0);
    }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [key]); 
  
  return success;
};

export const useDarkMode = () => {
  return useSecretCode(KONAMI_CODE);
};
