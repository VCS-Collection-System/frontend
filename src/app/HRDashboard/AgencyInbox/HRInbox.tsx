import React, { useState, useEffect } from 'react';
import { 
  Menu, 
  MenuContent, 
  MenuList, 
  MenuItem, 
  Card, 
  MenuGroup 
} from '@patternfly/react-core';
import { useHistory } from 'react-router-dom';
import CardWithTitle from '@app/Components/CardWithTitle';
import authaxios from "@app/utils/axiosInterceptor";
import { GenericResponseModal } from '@app/Components/FeedbackModals';
import { useKeycloak } from "@react-keycloak/web";
import { getEmployeeInfo } from '@app/utils/getEmployeeInfo';
import { 
  PAM_TASK_URL, 
  RESERVED_TASK_URL, 
  INPROGRESS_TASK_URL,
  ALL_TASK_URL
} from "@app/utils/constants"

const HRInbox: React.FunctionComponent = () => {
  const [activeItem, setActiveItem] = useState('0'); 
  const [inboxList, setInboxList] = useState([]);
  const [refreshData, setRefreshData] = useState(false)
  const [inProgressList, setInProgressList] = useState([]);
  const [isClaimModalOpen, setIsClaimModalOpen] = useState(false); 
  const history = useHistory();
  const { keycloak } = useKeycloak();
  const employee = getEmployeeInfo(keycloak);
  const reviewerID = employee.id;

  useEffect(() => {
    authaxios.get(RESERVED_TASK_URL + reviewerID)
    .then(resOne => {
      authaxios.get(INPROGRESS_TASK_URL + reviewerID)
      .then(resTwo => {
        setInProgressList(resOne.data["task-summary"].concat(resTwo.data["task-summary"]))
      })
    })

    authaxios.get(ALL_TASK_URL)
    .then(res => {
      setInboxList(res.data["task-summary"])
    })
  }, [refreshData, reviewerID]);

  const onSelect = (event, itemId) => {
    const itemIndex = parseInt(itemId.split(':')[1]);
    const taskUrl = PAM_TASK_URL + itemIndex + "/states"

    const headers = {
      'Content-Type': 'application/json'
    }

    if(itemId.includes('INPROGRESS')) {
      sessionStorage.setItem("approveId", "" + itemIndex);
      setActiveItem(itemId)
      history.push('/hrdashboard/review');
    }
    else {
      authaxios.put(taskUrl + "/claimed", {}, {
        headers: headers
      })
      .then(() => {
        sessionStorage.setItem("approveId", "" + itemIndex);
        setActiveItem(itemId)
        history.push('/hrdashboard/review');
      })
      .catch(() => {
        setIsClaimModalOpen(true)
      })
    }
  }

  const convertToDate = (dateString) => {
    const date = parseInt(dateString);
    const d = new Date(date);
    return  ("0"+(d.getMonth()+1)).slice(-2) + "-" + ("0" + d.getDate()).slice(-2) + "-" + d.getFullYear();
  }

  const closeAndRefresh = () => {
    setRefreshData(!refreshData); 
    setIsClaimModalOpen(false);
  }

  return (
    <Card isRounded>
      <GenericResponseModal
        isModalOpen={isClaimModalOpen}
        setIsModalOpen={closeAndRefresh}
        title="Task Was Claimed"
        message="Please select another one."
      />
      <CardWithTitle title="Inbox" info="Select an item below to review" />
      <Menu onSelect={onSelect} activeItemId={activeItem}>
        <MenuContent menuHeight="350px">
          <MenuGroup label="Your In-Progress Tasks">
            <MenuList id="in-progress-list" className="indentList">
              {
                inProgressList.length == 0 &&
                <MenuItem key="empty" itemId="empty" isDisabled>
                  [No In-Progress Tasks]
                </MenuItem>
              }
              {
                inProgressList.length > 0 &&
                  inProgressList.map(item => (
                    <MenuItem 
                    key={item["task-id"]} 
                    data-id={item["task-id"]}
                    description={"Submitted on: " + convertToDate(item["task-created-on"]["java.util.Date"])}
                    itemId={"INPROGRESS:" + item["task-id"]}>
                      {item["task-description"]}
                    </MenuItem>
                  ))
              }
            </MenuList>
          </MenuGroup>
          <MenuGroup label="All Available Tasks">
            <MenuList id="all-available-list" className="indentList">
              {
                inboxList.length == 0 &&
                <MenuItem key="empty" itemId="empty" isDisabled>
                  [No Available Tasks]
                </MenuItem>
              }
              {
                inboxList.length > 0 &&
                  inboxList.map(item => (
                    <MenuItem 
                    key={item["task-id"]}
                    data-id={item["task-id"]} 
                    description={"Submitted on: " + convertToDate(item["task-created-on"]["java.util.Date"])} 
                    itemId={"READY:" + item["task-id"]}>
                      {item["task-description"]}
                    </MenuItem>
                  ))
              }
            </MenuList>
          </MenuGroup>
        </MenuContent>
      </Menu>
    </Card>
  );
};

export { HRInbox };
