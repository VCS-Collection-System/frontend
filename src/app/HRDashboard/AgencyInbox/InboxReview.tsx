import React, { useEffect, useState } from 'react';
import { 
  Button,
  Card, 
  CardBody, 
  CardFooter,
  Flex,
  FlexItem,
  Grid,
  GridItem,
  Split,
} from '@patternfly/react-core';
import { CheckCircleIcon, ErrorCircleOIcon, ExternalLinkAltIcon } from '@patternfly/react-icons';
import CardWithTitle from '@app/Components/CardWithTitle';
import { useHistory } from 'react-router-dom';
import authaxios from "@app/utils/axiosInterceptor";
import ReviewInfoCard from '@app/Components/ReviewInfoCard';
import EmployeeInfoCard from '@app/Components/EmployeeInfoCard';
import { ResponseErrorModal } from '@app/Components/FeedbackModals';
import DeclineReasonModal from '@app/Components/DeclineReasonModal';
import { useKeycloak } from "@react-keycloak/web";
import { getEmployeeInfo } from '@app/utils/getEmployeeInfo';
import { PAM_TASK_URL, ATTACHMENT_URL } from "@app/utils/constants"

const PKG = "com.redhat.vcs.model.";
const VAX_DOCUMENT = PKG + "VaccineCardDocument";
const VAX_DOCUMENT_LIST = PKG + "VaccineDocumentList";
const VAX_BRAND = PKG + "VaccineBrand";
const VAX_EMPLOYEE = PKG + "Employee";
const VAX_ATTACHMENT = PKG + "Attachment";
const VAX_DOC_REVIEW = PKG + "DocumentReview";

const InboxReview: React.FunctionComponent = () => {
  const history = useHistory();
  const { keycloak } = useKeycloak();
  const employee = getEmployeeInfo(keycloak);
  const reviewerID = employee.id;

  const [isFormSubmitting, setIsFormSubmitting] = useState(false);
  const [isDeclineModalOpen, setIsDeclineModalOpen] = useState(false); 
  const [isErrorModalOpen, setIsErrorModalOpen] = useState(false);
  const [errorState, setErrorState] = useState(''); 
  const [doses, setDoses] = useState<Array<Record<string, string>>>([]);
  const [employeeId, setEmployeeId] = useState("");
  const [employeeName, setEmployeeName] = useState("");
  const [employeeDOB, setEmployeeDOB] = useState("");
  const [employeeEmail, setEmployeeEmail] = useState("");
  const [attachment, setAttachment] = useState("");
  const [attachmentType, setAttachmentType] = useState(""); 
  const [attachmentSize, setAttachmentSize] = useState(0); 

  useEffect(() => {
    const taskNumber = sessionStorage.getItem("approveId")
    const taskUrl = PAM_TASK_URL + taskNumber + "?withInputData=true&withOutputData=true&withAssignments=true"

    authaxios.get(taskUrl)
    .then(result => {
      let vaxDocument = null;
      if(result['data']['task-input-data']['document']) {
        vaxDocument = result['data']['task-input-data']['document'][VAX_DOCUMENT];

        if(vaxDocument != null) {
          const vaccineShotDate = vaxDocument['vaccineAdministrationDate'];
          const vaxDoses = [{
            vaccineBrand: vaxDocument['vaccineBrand'][VAX_BRAND],
            vaccineDate: vaccineShotDate['year'] + "-" + vaccineShotDate['monthValue'] + "-" + vaccineShotDate['dayOfMonth'],
            vaccineLotNumber: vaxDocument['lotNumber']
          }];
          setDoses(vaxDoses);
        }
      } else {
        const vaxDocuments = result['data']['task-input-data']['documentList'][VAX_DOCUMENT_LIST]['documents'];

        const vaxDoses = vaxDocuments.map((dose) => {
          dose = dose[VAX_DOCUMENT];
          const vaccineShotDate = dose['vaccineAdministrationDate'];
          return {
            vaccineBrand: dose['vaccineBrand'][VAX_BRAND],
            vaccineDate: vaccineShotDate['year'] + "-" + vaccineShotDate['monthValue'].toString().padStart(2, '0') + "-" + vaccineShotDate['dayOfMonth'].toString().padStart(2, '0'),
            vaccineLotNumber: dose['lotNumber']
          }
        });
        // Must sort because order of doses from server is not guaranteed to be in date order
        vaxDoses.sort((a, b) => {
          return a.vaccineDate.localeCompare(b.vaccineDate);
        });
        setDoses(vaxDoses);

        vaxDocument = vaxDocuments[vaxDocuments.length-1][VAX_DOCUMENT];
      }

      if(vaxDocument != null) {
        const employeeInfo = vaxDocument['employee'][VAX_EMPLOYEE]
        const dateOfBirth = employeeInfo['dateOfBirth']
        const resAttachment = vaxDocument['attachment'][VAX_ATTACHMENT]
        setAttachmentSize(resAttachment['size']);
        setEmployeeId(employeeInfo['employeeId'])
        setEmployeeName(employeeInfo['firstName'] + " " + employeeInfo['lastName'])
        if(dateOfBirth) {
          setEmployeeDOB(dateOfBirth['year'] + "-" + dateOfBirth['monthValue'] + "-" + dateOfBirth['dayOfMonth'])
        }
        setEmployeeEmail(employeeInfo['email']);

        authaxios.post(ATTACHMENT_URL, resAttachment, {
          responseType: 'arraybuffer'
        })
        .then(attachmentResult => {
          setAttachment(attachmentResult['data']);
          setAttachmentType(attachmentResult['headers']['content-type']);
        })
      }
    })
  }, []);

  const sendReview = (accepted, reason) => {
    setIsFormSubmitting(true)
    const taskNumber = sessionStorage.getItem("approveId")
    const taskUrl = PAM_TASK_URL + taskNumber + "/states/completed?auto-progress=true"
    const outcome = accepted ? "ACCEPTED" : "DECLINED"

    const acceptData = {
      "documentReview": {}
    }

    acceptData["documentReview"][VAX_DOC_REVIEW] = {
      "reviewerEmployeeId": reviewerID,
      "reviewOutcome": outcome,
      "reviewerNotes": reason,
      "rejectReason": reason
    }

    authaxios.put(taskUrl, acceptData, {
      headers: {
        'Accept': 'application/json',
        'Content-Type': 'application/json'
      }
    })
    .then(() => {
      history.push("/hrdashboard")
    })
    .catch(res => {
      if(res.response) {
        setErrorState(res.response.status);
        setIsErrorModalOpen(true);
      }
    }).finally(()=> {
      setIsDeclineModalOpen(false)
      setIsFormSubmitting(false)
    })
  }

  const releaseTask = () => {
    const taskNumber = sessionStorage.getItem("approveId")
    const taskUrl = PAM_TASK_URL + taskNumber + "/states/released"
    const headers = {
      'Content-Type': 'application/json'
    }

    authaxios.put(taskUrl, {}, {
      headers: headers
    })
    .then(() => {
      history.push("/hrdashboard")
    })
    .catch(() => {
      history.push("/hrdashboard")
    })
  }

  return (
    <Card isRounded>
      <ResponseErrorModal 
        isModalOpen={isErrorModalOpen} 
        setIsModalOpen={setIsErrorModalOpen} 
        errorState={errorState}
      />
      <DeclineReasonModal 
        isModalOpen={isDeclineModalOpen}
        isSendingReview={isFormSubmitting}
        setIsModalOpen={setIsDeclineModalOpen}
        sendReview={sendReview}
      />
      <CardWithTitle title="HR Review" info="Review Vaccination Record"/>
      <CardBody>
        <Grid hasGutter>
          <GridItem rowSpan={2} span={6}> 
            <Flex justifyContent={{ default: "justifyContentCenter" }}>
              <FlexItem>
                {
                  attachment && attachmentSize === 0 &&
                  "No image"
                }
                {
                  attachment && attachmentType != 'application/pdf' && attachmentSize > 0 &&
                  <img src={URL.createObjectURL(new Blob([attachment]))} /> 
                }
                {
                  attachment && attachmentType == 'application/pdf' && attachmentSize > 0 &&
                  (
                    <Button
                      icon={<ExternalLinkAltIcon />}
                      variant="link"
                      component="a"
                      href={URL.createObjectURL(new Blob([attachment], {type: 'application/pdf'}))}
                      target="_blank"
                    >
                      Vaccination Card PDF
                    </Button>
                  )
                }
              </FlexItem>
            </Flex>
          </GridItem>
          <GridItem span={6}> 
            <ReviewInfoCard
              reviewType="VAX"
              doses={doses}
            />
            <EmployeeInfoCard 
              employeeId={employeeId}
              employeeName={employeeName}
              employeeDOB={employeeDOB}
              employeeEmail={employeeEmail}
            />
          </GridItem>
        </Grid>
      </CardBody>
      <CardFooter>
        <Split hasGutter>
          <Button 
            icon={<CheckCircleIcon />}
            isDisabled={isFormSubmitting}
            isLoading={isFormSubmitting}
            onClick={() => sendReview(true, "")}
            variant="primary" 
          >
            Accept
          </Button>
          <Button 
            icon={<ErrorCircleOIcon />} 
            variant="danger" 
            onClick = {() => setIsDeclineModalOpen(true)}
          >
            Decline
          </Button>
          <Button 
            variant="plain" 
            onClick={() => releaseTask()}
          >
            Back
          </Button>
        </Split>
      </CardFooter>
    </Card>
  );
}

export { InboxReview };
