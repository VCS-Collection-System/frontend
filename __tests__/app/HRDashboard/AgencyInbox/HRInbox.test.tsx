import React from 'react';
import { HRInbox } from '../../../../src/app/HRDashboard/AgencyInbox/HRInbox';
import { act } from 'react-dom/test-utils';
import { RESERVED_TASK_URL, INPROGRESS_TASK_URL, ALL_TASK_URL, PAM_TASK_URL } from '../../../../src/app/utils/constants'
import { fireEvent, render, screen } from '@testing-library/react'
import reactRouterDom from 'react-router-dom'
import mockKeycloak from '../../../../__mocks__/keycloakMock'

jest.mock('@react-keycloak/web', () => ({
  useKeycloak: jest.fn( ()=>({keycloak: mockKeycloak}))
})) 

jest.mock("@app/utils/constants", () => ({
  KEYCLOAK_FIRST_NAME: "firstName",
  KEYCLOAK_LAST_NAME: "lastName",
  KEYCLOAK_EMAIL: "email",
  KEYCLOAK_COUNTRY: "attributes.agency[0]",
  KEYCLOAK_ID: "attributes.workforceid[0]",
  APPROVER_ROLES: ["mock-approver"],
  RESERVED_TASK_URL: "reserve-url/",
  INPROGRESS_TASK_URL: "inprogress-url/",
  ALL_TASK_URL: "task-url/",
  PAM_TASK_URL: "pam-url/"
}))

jest.mock('react-router-dom')
const mockPush = jest.fn()
reactRouterDom.useHistory = jest.fn().mockReturnValue({push: mockPush})
let reservedTasks = {"task-summary": [{'task-id': "mockID", "task-created-on": {"java.util.Date": "1508484583331"}, "task-description": "mockReserved"}]};
let inprogressTasks = {"task-summary": [{'task-id': "mockID", "task-created-on": {"java.util.Date": "1508484583331"}, "task-description": "mockProgress"}]};
let allTasks = {"task-summary": [{'task-id': "mockID", "task-created-on": {"java.util.Date": "1508484583331"}, "task-description": "mockTask"}]};

jest.mock('@app/utils/axiosInterceptor', () => ({
  get: (URL) => {
    let returnData; 
    if(URL === RESERVED_TASK_URL + "mockId"){
      returnData = {data: reservedTasks}
    }
    else if(URL === INPROGRESS_TASK_URL + "mockId"){
      returnData = {data: inprogressTasks}
    }
    else if(URL === ALL_TASK_URL){
      returnData = {data: allTasks}
    }

    else{
      returnData = {data: []};
    }

    return Promise.resolve(returnData)
  },
  put: (URL) => {
    let returnData; 

    if(URL === PAM_TASK_URL + "2" + "/states/claimed"){
      returnData = {data: allTasks}
    }
    else if(URL === PAM_TASK_URL + "3" + "/states/claimed"){
      return Promise.reject()
    }
    return Promise.resolve(returnData)
  }
}))

beforeEach(() => {
  reservedTasks = {"task-summary": [{'task-id': "0", "task-created-on": {"java.util.Date": "1508484583331"}, "task-description": "mockReserved"}]};
  inprogressTasks = {"task-summary": [{'task-id': "1", "task-created-on": {"java.util.Date": "1508484583331"}, "task-description": "mockProgress"}]};
  allTasks = {"task-summary": [{'task-id': "2", "task-created-on": {"java.util.Date": "1508484583331"}, "task-description": "mockTask"},
                                {'task-id': "3", "task-created-on": {"java.util.Date": "1508484583331"}, "task-description": "mockTask2"} ]};
})

afterEach(() => {
  sessionStorage.clear(); 
})

describe('<HRInbox />', ()=> {
  it('should be defined', () => {
    expect(HRInbox).toBeDefined()
  })

  it("should match existing snapshot", async () => {
    await act(async () => {
      const { asFragment } = render(<HRInbox />)
      expect(asFragment()).toMatchSnapshot()
    })
  })

  it("should display empty list when no tasks", async () => {
    reservedTasks = {"task-summary": []};
    inprogressTasks = {"task-summary": []};
    allTasks = {"task-summary": []};
    await act(async () => {
      render(<HRInbox />)
    })
    expect(screen.getByText('[No In-Progress Tasks]')).toBeTruthy(); 
    expect(screen.getByText('[No Available Tasks]')).toBeTruthy(); 
  })

  it("should display list with tasks", async () => {
    await act(async () => {
      render(<HRInbox />)
    })
    expect(screen.getByRole('button', {name: /mockReserved Submitted on: 10-20-2017/i})).toBeTruthy();
    expect(screen.getByRole('button', {name: /mockProgress Submitted on: 10-20-2017/i})).toBeTruthy();
    expect(screen.getByRole('button', {name: /mockTask Submitted on: 10-20-2017/i})).toBeTruthy();
    expect(screen.getByRole('button', {name: /mockTask2 Submitted on: 10-20-2017/i})).toBeTruthy();
  })

  it("should allow in progress tasks to be clickable", async () => {
    await act(async () => {
      render(<HRInbox />)
    })
    const progressItem = screen.getByRole('button', {name: /mockProgress Submitted on: 10-20-2017/i});
    fireEvent.click(progressItem);
    expect(sessionStorage.getItem("approveId")).toEqual("1")
    expect(mockPush).toBeCalledWith('/hrdashboard/review')
  })

  it("should allow regular tasks to be clickable", async () => {
    await act(async () => {
      render(<HRInbox />)
    })
    const taskItem = screen.getByRole('button', {name: /mockTask Submitted on: 10-20-2017/i});
    await act(async () => {
      fireEvent.click(taskItem);
    })
    expect(sessionStorage.getItem("approveId")).toEqual("2")
    expect(mockPush).toBeCalledWith('/hrdashboard/review')
  })

  it("should display modal when task already claimed", async () => {
    await act(async () => {
      render(<HRInbox />)
    })
    const taskItem = screen.getByRole('button', {name: /mockTask2 Submitted on: 10-20-2017/i});
    await act(async () => {
      fireEvent.click(taskItem);
    })
    expect(screen.getByText('Task Was Claimed')).toBeTruthy(); 

    const closeModal = screen.getAllByRole('button', {name: /Close/i})[1];
    await act(async () => {
      fireEvent.click(closeModal);
    })

    expect(screen.queryAllByText('Task Was Claimed')).toEqual([]);
  })
})