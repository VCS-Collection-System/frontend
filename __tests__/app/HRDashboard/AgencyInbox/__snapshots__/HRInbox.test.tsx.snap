// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HRInbox /> should match existing snapshot 1`] = `
<DocumentFragment>
  <article
    class="pf-c-card pf-m-rounded"
    data-ouia-component-id="OUIA-Generated-Card-1"
    data-ouia-component-type="PF4/Card"
    data-ouia-safe="true"
    id=""
  >
    <div
      class="pf-c-card__title"
    >
      <p
        class=""
        data-pf-content="true"
      >
        Inbox
      </p>
      <div
        class="pf-c-content"
      >
        <small
          class=""
          data-pf-content="true"
        >
          Select an item below to review
        </small>
      </div>
    </div>
    <div
      aria-label="Global"
      class="pf-c-menu"
      data-ouia-component-id="OUIA-Generated-Menu-1"
      data-ouia-component-type="PF4/Menu"
      data-ouia-safe="true"
    >
      <div
        class="pf-c-menu__content"
        style="--pf-c-menu__content--Height: 350px;"
      >
        <section
          class="pf-c-menu__group"
        >
          <h1
            class="pf-c-menu__group-title"
            id=""
          >
            Your In-Progress Tasks
          </h1>
          <ul
            class="pf-c-menu__list indentList"
            id="in-progress-list"
          >
            <li
              class="pf-c-menu__list-item pf-m-disabled"
              tabindex="-1"
            >
              <button
                aria-current="false"
                class="pf-c-menu__item"
                disabled=""
                tabindex="0"
                type="button"
              >
                <span
                  class="pf-c-menu__item-main"
                >
                  <span
                    class="pf-c-menu__item-text"
                  >
                    [No In-Progress Tasks]
                  </span>
                </span>
              </button>
            </li>
          </ul>
        </section>
        <section
          class="pf-c-menu__group"
        >
          <h1
            class="pf-c-menu__group-title"
            id=""
          >
            All Available Tasks
          </h1>
          <ul
            class="pf-c-menu__list indentList"
            id="all-available-list"
          >
            <li
              class="pf-c-menu__list-item pf-m-disabled"
              tabindex="-1"
            >
              <button
                aria-current="false"
                class="pf-c-menu__item"
                disabled=""
                tabindex="-1"
                type="button"
              >
                <span
                  class="pf-c-menu__item-main"
                >
                  <span
                    class="pf-c-menu__item-text"
                  >
                    [No Available Tasks]
                  </span>
                </span>
              </button>
            </li>
          </ul>
        </section>
      </div>
    </div>
  </article>
</DocumentFragment>
`;
